### ISSUE TITLE
false positive advice to remove annotation used library

### ISSUE THREAD

**Issue by xenoterracide** [autonomousapps/dependency-analysis-gradle-plugin#1274]
**Build scan link**

https://gradle.com/s/532wov7wlgtsg

**Plugin version**

2.1.1

**Gradle version**

```
------------------------------------------------------------
Gradle 8.10.2
------------------------------------------------------------

Build time:    2024-09-23 21:28:39 UTC
Revision:      415adb9e06a516c44b391edff552fd42139443f7

Kotlin:        1.9.24
Groovy:        3.0.22
Ant:           Apache Ant(TM) version 1.10.14 compiled on August 16 2023
Launcher JVM:  21.0.4 (Eclipse Adoptium 21.0.4+7-LTS)
Daemon JVM:    /home/xeno/.asdf/installs/java/temurin-21.0.4+7.0.LTS (no JDK specified, using current Java home)
OS:            Linux 6.6.47-1-MANJARO amd64
```

**Describe the bug**

That library has usages, it's used on `package` and in `module-info` (yes, I know on the latter not supported)

```  
  Advice for :security-model
  Unused dependencies which should be removed:
    implementation(libs.spring.modulith.api)
```

**To Reproduce**

speculatively, a library is used only with annotations that are annotating a `package-info` class

```java
/**
 * The {@link com.xenoterracide.model.security.user.User} aggregate.
 */
@DomainLayer
@NamedInterface
package com.xenoterracide.model.security.user;

import org.jmolecules.architecture.layered.DomainLayer;
import org.springframework.modulith.NamedInterface;
```

**Expected behavior**

no error. implementation is the correct scope.

**Additional context**

wasn't reported in 1.x

**Comment by pexa-dnugent** [comment]
I see the same thing with Kotlin-reflect, although the advice here is to make it runtime only. The problem is that if I do that, I get compiler warnings.

**Comment by autonomousapps** [comment]
Please test the latest version, there have been a lot of changes/fixes around this lately.

**Comment by tristankretzer** [comment]
I have the same issue with version 2.5.0. I created a minimal repository to reproduce this issue: https://github.com/tristankretzer/autonomousapps-dependency-analysis-gradle-plugin-issue-1274.

**Comment by autonomousapps** [comment]
Thanks!

**Comment by autonomousapps** [comment]
@tristankretzer I think that your issue may be different from the original. Nonetheless, I believe it is resolved with https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1723.

I plan to work on the original issue soon. I think that issue is specific to annotations on packages. Yours was an issue with KMP libs and typealiases, and specifically the kotlin-test lib, which has this:
```
@file:JvmPackageName("kotlin.test.junit5.annotations")
package kotlin.test
```
Why? Who knows.

**Comment by autonomousapps** [comment]
@xenoterracide I know it's been a while, so you may not be able to respond easily. I have tried to reproduce your situation and this is the output:
```
Advice for :consumer
Dependencies which could be compile-only:
  compileOnly project(':annotations') (was implementation)
```
This is my annotation:
```
package com.annotations;

import java.lang.annotation.*;

@Target(ElementType.PACKAGE)
public @interface Magic {}
```
and this is my `package-info.java`:
```
@Magic
package com.consumer;

import com.annotations.Magic;
```
I've tried this with `SOURCE`, `CLASS`, and `RUNTIME` retention policies and the result is always the same.

tl;dr I don't think your issue is reproducible any longer. I've added a test case for this at https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1723.
